# Actividad Evaluable 2 - Manejo de datos con Pandas


---


Elizabeth Jauregui Zarate
A01253381

# Data management using Pandas
Data management is a crucial component to statistical analysis and data science work.

This notebook will show you how to import, view, undertand, and manage your data using the Pandas data processing library, i.e., the notebook will demonstrates how to read a dataset into Python, and obtain a basic understanding of its content.

Note that Python by itself is a general-purpose programming language and does not provide high-level data processing capabilities. The Pandas library was developed to meet this need. Pandas is the most popular Python library for data manipulation, and we will use it extensively in this course. Pandas provides high-performance, easy-to-use data structures and data analysis tools.

The main data structure that Pandas works with is called a Data Frame. This is a two-dimensional table of data in which the rows typically represent cases and the columns represent variables (e.g. data used in this tutorial). Pandas also has a one-dimensional data structure called a Series that we will encounter when accesing a single column of a Data Frame.

Pandas has a variety of functions named read_xxx for reading data in different formats. Right now we will focus on reading csv files, which stands for comma-separated values. However the other file formats include excel, json, and sql.

There are many other options to read_csv that are very useful. For example, you would use the option sep='\t' instead of the default sep=',' if the fields of your data file are delimited by tabs instead of commas. See here for the full documentation for read_csv.

# Importing libraries

In [ ]:
# Import the packages that we will be using
import pandas as pd

# Importing data

In [ ]:
# url string that hosts our .csv file
url = "cartwheel.csv"

# Read the .csv file and store it as a pandas Data Frame
df = pd.read_csv(url)
#type(df)

# Exploring the content of the data set
Use the shape method to determine the numbers of rows and columns in a data frame. This can be used to confirm that we have actually obtained the data the we are expecting.

In [ ]:
# Print the number of rows
num_rows = len(df.index)
print(f"Number of rows: {num_rows}")

Number of rows: 52


In [ ]:
# Print the number of columns
num_columns = len(df.columns)
print(f"Number of columns: {num_columns}")

Number of columns: 12


In [ ]:
# If we want to show the entire data frame we would simply write the following:
print(df)

    ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0    1  56.0      F            1       Y             1   62.00      61.0   
1    2  26.0      F            1       Y             1   62.00      60.0   
2    3  33.0      F            1       Y             1   66.00      64.0   
3    4  39.0      F            1       N             0   64.00      63.0   
4    5  27.0      M            2       N             0   73.00      75.0   
5    6  24.0      M            2       N             0   75.00      71.0   
6    7  28.0      M            2       N             0   75.00      76.0   
7    8  22.0      F            1       N             0   65.00      62.0   
8    9  29.0      M            2       Y             1   74.00      73.0   
9   10  33.0      F            1       Y             1   63.00      60.0   
10  11  30.0      M            2       Y             1   69.50      66.0   
11  12  28.0      F            1       Y             1   62.75      58.0   
12  13  25.0

In [ ]:
#Now, use the the head() function to show the first 5 rows of our data frame
df.head()

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score  
0          79        Y            1.0      7  
1          70        Y            1.0      8  
2          85        Y            1.0      7  
3          87        Y            1.0     10  
4          72        N            0.0      4

In [ ]:
#Also, you can use the the tail() function to show the last 5 rows of our data frame
df.tail()

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
47  48  24.0      M            2       N             0    79.5      75.0   
48  49  28.0      M            2       N             0    77.8      76.0   
49  50  30.0      F            1       N             0    74.6       NaN   
50  51   NaN      M            2       N             0    71.0      70.0   
51  52  27.0      M            2       N             0     NaN      71.5   

    CWDistance Complete  CompleteGroup  Score  
47          82        N            0.0      8  
48          99        Y            1.0      9  
49          71        Y            1.0      9  
50         101        Y            NaN      8  
51         103        Y            1.0     10

In [ ]:
#The columns in a Pandas data frame have names, to see the names, use the columns method:

#To gather more information regarding the data, we can view the column names with the following function:
df.columns

Index(['ID', 'Age', 'Gender', 'GenderGroup', 'Glasses', 'GlassesGroup',
       'Height', 'Wingspan', 'CWDistance', 'Complete', 'CompleteGroup',
       'Score'],
      dtype='object')

In [ ]:
#Be aware that every variable in a Pandas data frame has a data type. There are many different data types, but most commonly you will encounter floating point values (real numbers), integers, strings (text), and date/time values. When Pandas reads a text/csv file, it guesses the data types based on what it sees in the first few rows of the data file. Usually it selects an appropriate type, but occasionally it does not. To confirm that the data types are consistent with what the variables represent, inspect the dtypes attribute of the data frame.
df.dtypes

ID                 int64
Age              float64
Gender            object
GenderGroup        int64
Glasses           object
GlassesGroup       int64
Height           float64
Wingspan         float64
CWDistance         int64
Complete          object
CompleteGroup    float64
Score              int64
dtype: object

In [ ]:
#Summary statistics, which include things like the mean, min, and max of the data, can be useful to get a feel for how large some of the variables are and what variables may be the most important.
#Summary statistics for the quantitative variables:
df.describe()

ID        Age  GenderGroup  GlassesGroup     Height   Wingspan  \
count  52.000000  51.000000    52.000000     52.000000  51.000000  51.000000   
mean   26.500000  28.411765     1.500000      0.500000  68.971569  67.313725   
std    15.154757   5.755611     0.504878      0.504878   5.303812   5.624021   
min     1.000000  22.000000     1.000000      0.000000  61.500000  57.500000   
25%    13.750000  25.000000     1.000000      0.000000  64.500000  63.000000   
50%    26.500000  27.000000     1.500000      0.500000  69.000000  66.000000   
75%    39.250000  30.000000     2.000000      1.000000  73.000000  72.000000   
max    52.000000  56.000000     2.000000      1.000000  79.500000  76.000000   

       CWDistance  CompleteGroup      Score  
count   52.000000      51.000000  52.000000  
mean    85.576923       0.843137   7.173077  
std     14.353173       0.367290   2.211566  
min     63.000000       0.000000   2.000000  
25%     72.000000       1.000000   6.000000  
50%     85.000000       1.000000   8.000000  
75%     96.500000       1.000000   9.000000  
max    115.000000       1.000000  10.000000

In [ ]:
# Drop observations with NaN values
df.Age.dropna().describe()

count    51.000000
mean     28.411765
std       5.755611
min      22.000000
25%      25.000000
50%      27.000000
75%      30.000000
max      56.000000
Name: Age, dtype: float64

In [ ]:
# Drop observations with NaN values
df.Wingspan.dropna().describe()

count    51.000000
mean     67.313725
std       5.624021
min      57.500000
25%      63.000000
50%      66.000000
75%      72.000000
max      76.000000
Name: Wingspan, dtype: float64

It is also possible to get statistics on the entire data frame or a column as follows:
- df.mean() Returns the mean of all columns
- df.corr() Returns the correlation between columns in a data frame
- df.count() Returns the number of non-null values in each data frame column
- df.max() Returns the highest value in each column
- df.min() Returns the lowest value in each column
- df.median() Returns the median of each column
- df.std() Returns the standard deviation of each column

In [ ]:
numeric_df = df.select_dtypes(include=['number'])
numeric_df.mean()

ID               26.500000
Age              28.411765
GenderGroup       1.500000
GlassesGroup      0.500000
Height           68.971569
Wingspan         67.313725
CWDistance       85.576923
CompleteGroup     0.843137
Score             7.173077
dtype: float64

In [ ]:
# Calculate correlations for numeric columns only
correlation_matrix = df.corr(numeric_only=True)

# Display the correlation matrix
print(correlation_matrix)

                     ID       Age  GenderGroup  GlassesGroup    Height  \
ID             1.000000 -0.139088     0.066630     -0.184513  0.313570   
Age           -0.139088  1.000000    -0.263563      0.153441 -0.321105   
GenderGroup    0.066630 -0.263563     1.000000     -0.230769  0.731284   
GlassesGroup  -0.184513  0.153441    -0.230769      1.000000 -0.251782   
Height         0.313570 -0.321105     0.731284     -0.251782  1.000000   
Wingspan       0.272402 -0.250976     0.763129     -0.212414  0.941516   
CWDistance     0.120251 -0.030501     0.265168     -0.075762  0.315947   
CompleteGroup  0.110208  0.229843    -0.224166      0.116313 -0.242616   
Score          0.223190  0.317358    -0.272192     -0.149266 -0.042706   

               Wingspan  CWDistance  CompleteGroup     Score  
ID             0.272402    0.120251       0.110208  0.223190  
Age           -0.250976   -0.030501       0.229843  0.317358  
GenderGroup    0.763129    0.265168      -0.224166 -0.272192  
Glasses

In [ ]:
df.count()

ID               52
Age              51
Gender           52
GenderGroup      52
Glasses          52
GlassesGroup     52
Height           51
Wingspan         51
CWDistance       52
Complete         52
CompleteGroup    51
Score            52
dtype: int64

In [ ]:
numeric_df = df.select_dtypes(include=['number'])
numeric_df.max()

ID                52.0
Age               56.0
GenderGroup        2.0
GlassesGroup       1.0
Height            79.5
Wingspan          76.0
CWDistance       115.0
CompleteGroup      1.0
Score             10.0
dtype: float64

In [ ]:
numeric_df = df.select_dtypes(include=['number'])
numeric_df.min()

ID                1.0
Age              22.0
GenderGroup       1.0
GlassesGroup      0.0
Height           61.5
Wingspan         57.5
CWDistance       63.0
CompleteGroup     0.0
Score             2.0
dtype: float64

In [ ]:
numeric_df = df.select_dtypes(include=['number'])
numeric_df.median()

ID               26.5
Age              27.0
GenderGroup       1.5
GlassesGroup      0.5
Height           69.0
Wingspan         66.0
CWDistance       85.0
CompleteGroup     1.0
Score             8.0
dtype: float64

In [ ]:
numeric_df = df.select_dtypes(include=['number'])
numeric_df.std()

ID               15.154757
Age               5.755611
GenderGroup       0.504878
GlassesGroup      0.504878
Height            5.303812
Wingspan          5.624021
CWDistance       14.353173
CompleteGroup     0.367290
Score             2.211566
dtype: float64

# How to write a data frame to a File
To save a file with your data simply use the to_csv attribute

In [ ]:
df.to_csv('myDataFrame.csv')

# Rename columns
To change the name of a colum use the rename attribute

In [ ]:
df = df.rename(columns={"Age": "Edad"})

df.head()

ID  Edad Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score  
0          79        Y            1.0      7  
1          70        Y            1.0      8  
2          85        Y            1.0      7  
3          87        Y            1.0     10  
4          72        N            0.0      4

In [ ]:
# Back to the original name
df = df.rename(columns={"Edad": "Age"})

df.head()

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score  
0          79        Y            1.0      7  
1          70        Y            1.0      8  
2          85        Y            1.0      7  
3          87        Y            1.0     10  
4          72        N            0.0      4

# Selection of colums
As discussed above, a Pandas data frame is a rectangular data table, in which the rows represent observations or samples and the columns represent variables. One common manipulation of a data frame is to extract the data for one case or for one variable. There are several ways to do this, as shown below.



In [ ]:
#To extract all the values for one column (variable), use one of the following alternatives.
a = df.Age
#b = df["Age"]
#c = df.loc[:, "Age"]
#d = df.iloc[:, 1]

print(a)

#df[["Gender", "GenderGroup"]]

0     56.0
1     26.0
2     33.0
3     39.0
4     27.0
5     24.0
6     28.0
7     22.0
8     29.0
9     33.0
10    30.0
11    28.0
12    25.0
13    23.0
14    31.0
15    26.0
16    26.0
17    27.0
18    23.0
19    24.0
20    23.0
21    29.0
22    25.0
23    26.0
24    23.0
25    28.0
26    24.0
27    25.0
28    32.0
29    38.0
30    27.0
31    33.0
32    38.0
33    27.0
34    24.0
35    27.0
36    25.0
37    26.0
38    31.0
39    30.0
40    23.0
41    26.0
42    28.0
43    26.0
44    30.0
45    39.0
46    27.0
47    24.0
48    28.0
49    30.0
50     NaN
51    27.0
Name: Age, dtype: float64


# Slicing a data set
The attibute .loc() uses labels/column names, in specific, it takes two single/list/range operator separated by ',', the first one indicates the rows and the second one indicates columns.

In [ ]:
# Return all observations of CWDistance
df.loc[:,"CWDistance"]

# Return a subset of observations of CWDistance
df.loc[:9, "CWDistance"]

# Select all rows for multiple columns, ["Gender", "GenderGroup"]
df.loc[:,["Gender", "GenderGroup"]]

# Select multiple columns, ["Gender", "GenderGroup"]me
keep = ['Gender', 'GenderGroup']
df_gender = df[keep]

# Select few rows for multiple columns, ["CWDistance", "Height", "Wingspan"]
df.loc[4:9, ["CWDistance", "Height", "Wingspan"]]

# Select range of rows for all columns
df.loc[10:15,:]

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
10  11  30.0      M            2       Y             1   69.50      66.0   
11  12  28.0      F            1       Y             1   62.75      58.0   
12  13  25.0      F            1       Y             1   65.00      64.5   
13  14  23.0      F            1       N             0   61.50      57.5   
14  15  31.0      M            2       Y             1   73.00      74.0   
15  16  26.0      M            2       Y             1   71.00      72.0   

    CWDistance Complete  CompleteGroup  Score  
10          96        Y            1.0      6  
11          79        Y            1.0     10  
12          92        Y            1.0      6  
13          66        Y            1.0      4  
14          72        Y            1.0      9  
15         115        Y            1.0      6

The attribute iloc() is an integer based slicing.

In [ ]:
# .
df.iloc[:, :4]

# .
df.iloc[:4, :]

# .
df.iloc[:, 3:7]

# .
df.iloc[4:8, 2:4]

# This is incorrect:
#df.iloc[1:5, ["Gender", "GenderGroup"]]

Gender  GenderGroup
4      M            2
5      M            2
6      M            2
7      F            1

# Get unique existing values
List unique values in the one of the columns

In [ ]:
# List unique values in the df['Gender'] column:
df.Gender.unique()

array(['F', 'M'], dtype=object)

In [ ]:
# Lets explore df["GenderGroup] as well:
df.GenderGroup.unique()

array([1, 2])

# Filter, Sort and Groupby
With Filter you can use different conditions to filter columns. For example, df[df[year] > 1984] would give you only the column year is greater than 1984. You can use & (and) or | (or) to add different conditions to your filtering. This is also called boolean filtering.

In [ ]:
df[df["Height"] >= 70]

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
4    5  27.0      M            2       N             0    73.0      75.0   
5    6  24.0      M            2       N             0    75.0      71.0   
6    7  28.0      M            2       N             0    75.0      76.0   
8    9  29.0      M            2       Y             1    74.0      73.0   
14  15  31.0      M            2       Y             1    73.0      74.0   
15  16  26.0      M            2       Y             1    71.0      72.0   
18  19  23.0      M            2       Y             1    70.0      69.0   
21  22  29.0      M            2       N             0    71.0      70.0   
22  23  25.0      M            2       N             0    70.0      68.0   
25  26  28.0      M            2       N             0    75.0      76.0   
26  27  24.0      M            2       N             0    78.4      71.0   
27  28  25.0      M            2       Y             1    76.0      73.0   
33  34  27.0      M            2       N             0    77.0      75.0   
38  39  31.0      M            2       Y             1    73.0      74.0   
40  41  23.0      F            1       N             0    70.4      71.0   
41  42  26.0      M            2       Y             1    73.5      72.0   
42  43  28.0      F            1       Y             1    72.5      72.0   
43  44  26.0      F            1       Y             1    72.0      72.0   
46  47  27.0      M            2       N             0    78.0      75.0   
47  48  24.0      M            2       N             0    79.5      75.0   
48  49  28.0      M            2       N             0    77.8      76.0   
49  50  30.0      F            1       N             0    74.6       NaN   
50  51   NaN      M            2       N             0    71.0      70.0   

    CWDistance Complete  CompleteGroup  Score  
4           72        N            0.0      4  
5           81        N            0.0      3  
6          107        Y            1.0     10  
8          106        N            0.0      5  
14          72        Y            1.0      9  
15         115        Y            1.0      6  
18          64        Y            1.0      3  
21         101        Y            1.0      8  
22          82        Y            1.0      4  
25         111        Y            1.0     10  
26          92        Y            1.0      7  
27         107        Y            1.0      8  
33         100        Y            1.0      8  
38          72        Y            1.0      9  
40          66        Y            1.0      4  
41         115        Y            1.0      6  
42          81        Y            1.0     10  
43          92        Y            1.0      8  
46          72        N            0.0      7  
47          82        N            0.0      8  
48          99        Y            1.0      9  
49          71        Y            1.0      9  
50         101        Y            NaN      8

With Sort is possible to sort values in a certain column in an ascending order using df.sort_values("ColumnName") or in descending order using df.sort_values(ColumnName, ascending=False).

Furthermore, it’s possible to sort values by Column1Name in ascending order then Column2Name in descending order by using df.sort_values([Column1Name,Column2Name],ascending=[True,False])

df.sort_values("Height")

# df.sort_values("Height",ascending=False)

In [ ]:
df.sort_values("Height",ascending=False)

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
47  48  24.0      M            2       N             0   79.50      75.0   
26  27  24.0      M            2       N             0   78.40      71.0   
46  47  27.0      M            2       N             0   78.00      75.0   
48  49  28.0      M            2       N             0   77.80      76.0   
33  34  27.0      M            2       N             0   77.00      75.0   
27  28  25.0      M            2       Y             1   76.00      73.0   
25  26  28.0      M            2       N             0   75.00      76.0   
5    6  24.0      M            2       N             0   75.00      71.0   
6    7  28.0      M            2       N             0   75.00      76.0   
49  50  30.0      F            1       N             0   74.60       NaN   
8    9  29.0      M            2       Y             1   74.00      73.0   
41  42  26.0      M            2       Y             1   73.50      72.0   
14  15  31.0      M            2       Y             1   73.00      74.0   
38  39  31.0      M            2       Y             1   73.00      74.0   
4    5  27.0      M            2       N             0   73.00      75.0   
42  43  28.0      F            1       Y             1   72.50      72.0   
43  44  26.0      F            1       Y             1   72.00      72.0   
15  16  26.0      M            2       Y             1   71.00      72.0   
21  22  29.0      M            2       N             0   71.00      70.0   
50  51   NaN      M            2       N             0   71.00      70.0   
40  41  23.0      F            1       N             0   70.40      71.0   
18  19  23.0      M            2       Y             1   70.00      69.0   
22  23  25.0      M            2       N             0   70.00      68.0   
39  40  30.0      M            2       Y             1   69.50      66.0   
10  11  30.0      M            2       Y             1   69.50      66.0   
20  21  23.0      M            2       Y             1   69.00      67.0   
23  24  26.0      M            2       N             0   69.00      71.0   
19  20  24.0      F            1       Y             1   68.00      66.0   
35  36  27.0      M            2       N             0   68.00      66.0   
34  35  24.0      F            1       N             0   67.80      62.0   
17  18  27.0      M            2       N             0   66.00      66.0   
44  45  30.0      F            1       Y             1   66.00      64.0   
2    3  33.0      F            1       Y             1   66.00      64.0   
31  32  33.0      F            1       Y             1   65.30      64.0   
12  13  25.0      F            1       Y             1   65.00      64.5   
24  25  23.0      F            1       Y             1   65.00      63.0   
36  37  25.0      F            1       Y             1   65.00      64.5   
7    8  22.0      F            1       N             0   65.00      62.0   
45  46  39.0      F            1       N             0   64.00      63.0   
3    4  39.0      F            1       N             0   64.00      63.0   
32  33  38.0      F            1       N             0   64.00      63.0   
28  29  32.0      F            1       Y             1   63.00      60.0   
9   10  33.0      F            1       Y             1   63.00      60.0   
11  12  28.0      F            1       Y             1   62.75      58.0   
30  31  27.0      F            1       Y             1   62.00      60.0   
0    1  56.0      F            1       Y             1   62.00      61.0   
1    2  26.0      F            1       Y             1   62.00      60.0   
16  17  26.0      F            1       N             0   61.50      59.5   
13  14  23.0      F            1       N             0   61.50      57.5   
29  30  38.0      F            1       Y             1   61.50      61.0   
37  38  26.0      F            1       N             0   61.50      59.5   
51  52  27.0      M            2       N            

The attribute Groupby involves splitting the data into groups based on some criteria, applying a function to each group independently and combining the results into a data structure. df.groupby(col) returns a groupby object for values from one column while df.groupby([col1,col2]) returns a groupby object for values from multiple columns.



In [ ]:
df.groupby(['Gender'])

In [ ]:
#Size of each group

df.groupby(['Gender']).size()

df.groupby(['Gender','GenderGroup']).size()

Gender  GenderGroup
F       1              26
M       2              26
dtype: int64

# Data Cleaning: handle with missing data
Before getting started to work with your data, it's a good practice to observe it thoroughly to identify missing values and handle them accordingly.

When reading a dataset using Pandas, there is a set of values including 'NA', 'NULL', and 'NaN' that are taken by default to represent a missing value. The full list of default missing value codes is in the 'read_csv' documentation here. This document also explains how to change the way that 'read_csv' decides whether a variable's value is missing.

Pandas has functions called isnull and notnull that can be used to identify where the missing and non-missing values are located in a data frame.

In [ ]:
#Below we use these functions to count the number of missing and non-missing values in each variable of the datasetr:
df.isnull().sum()

ID               0
Age              1
Gender           0
GenderGroup      0
Glasses          0
GlassesGroup     0
Height           1
Wingspan         1
CWDistance       0
Complete         0
CompleteGroup    1
Score            0
dtype: int64

In [ ]:
#Unfortunately, our output indicates that some of our columns contain missing values so we are no able to continue on doing analysis with those colums:
df.notnull().sum()

ID               52
Age              51
Gender           52
GenderGroup      52
Glasses          52
GlassesGroup     52
Height           51
Wingspan         51
CWDistance       52
Complete         52
CompleteGroup    51
Score            52
dtype: int64

In [ ]:
# Now we use these functions to count the number of missing and non-missing values in a single variable in the dataset

print( df.Height.notnull().sum() )

print( pd.isnull(df.Height).sum() )

51
1


In [ ]:
# Extract all non-missing values of one of the columns into a new variable
x = df.Age.dropna().describe()
x.describe()

count     8.000000
mean     30.645922
std      16.044470
min       5.755611
25%      24.250000
50%      27.705882
75%      35.250000
max      56.000000
Name: Age, dtype: float64

#Add and eliminate columns
In some cases it is useful to create or eiminate new columns

In [ ]:
df.head()
# Add a new column with new data

# Create a column data
NewColumnData = df.Age/df.Age

# Insert that column in the data frame
df.insert(12, "ColumnInserted", NewColumnData, True)

df.head()

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score  ColumnInserted  
0          79        Y            1.0      7             1.0  
1          70        Y            1.0      8             1.0  
2          85        Y            1.0      7             1.0  
3          87        Y            1.0     10             1.0  
4          72        N            0.0      4             1.0

In [ ]:
# Eliminate inserted column
#df.drop("ColumnInserted", axis=1, inplace = True)
#df.drop(columns=['ColumnInserted'], inplace = True)

# Remove three columns as index base
df.drop(df.columns[[12]], axis = 1, inplace = True)

df.head()

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score  
0          79        Y            1.0      7  
1          70        Y            1.0      8  
2          85        Y            1.0      7  
3          87        Y            1.0     10  
4          72        N            0.0      4

In [ ]:
# Add new column derived from existing columns

# The new column is a function of another column
df["AgeInMonths"] = df["Age"] * 12

df.head()

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score  AgeInMonths  
0          79        Y            1.0      7        672.0  
1          70        Y            1.0      8        312.0  
2          85        Y            1.0      7        396.0  
3          87        Y            1.0     10        468.0  
4          72        N            0.0      4        324.0

In [ ]:
# Eliminate inserted column
df.drop("AgeInMonths", axis=1, inplace = True)
df.head()

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score  
0          79        Y            1.0      7  
1          70        Y            1.0      8  
2          85        Y            1.0      7  
3          87        Y            1.0     10  
4          72        N            0.0      4

In [ ]:
# Add a new column with text labels reflecting the code's meaning

df["GenderGroupNew"] = df.GenderGroup.replace({1: "Female", 2: "Male"})

# Show the first 5 rows of the created data frame
df.head()

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score GenderGroupNew  
0          79        Y            1.0      7         Female  
1          70        Y            1.0      8         Female  
2          85        Y            1.0      7         Female  
3          87        Y            1.0     10         Female  
4          72        N            0.0      4           Male

In [ ]:
# Eliminate inserted column
df.drop("GenderGroupNew", axis=1, inplace = True)
##df.drop(['GenderGroupNew'],vaxis='columns',vinplace=True)
df.head()

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score  
0          79        Y            1.0      7  
1          70        Y            1.0      8  
2          85        Y            1.0      7  
3          87        Y            1.0     10  
4          72        N            0.0      4

In [ ]:
# Add a new column with strata based on these cut points

# Create a column data
NewColumnData = df.Age/df.Age

# Insert that column in the data frame
df.insert(1, "ColumnStrata", NewColumnData, True)

df["ColumnStrata"] = pd.cut(df.Height, [60., 63., 66., 69., 72., 75., 78.])

# Show the first 5 rows of the created data frame
df.head()

ID  ColumnStrata   Age Gender  GenderGroup Glasses  GlassesGroup  Height  \
0   1  (60.0, 63.0]  56.0      F            1       Y             1    62.0   
1   2  (60.0, 63.0]  26.0      F            1       Y             1    62.0   
2   3  (63.0, 66.0]  33.0      F            1       Y             1    66.0   
3   4  (63.0, 66.0]  39.0      F            1       N             0    64.0   
4   5  (72.0, 75.0]  27.0      M            2       N             0    73.0   

   Wingspan  CWDistance Complete  CompleteGroup  Score  
0      61.0          79        Y            1.0      7  
1      60.0          70        Y            1.0      8  
2      64.0          85        Y            1.0      7  
3      63.0          87        Y            1.0     10  
4      75.0          72        N            0.0      4

In [ ]:
# Eliminate inserted column
df.drop("ColumnStrata", axis=1, inplace = True)
df.head()

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score  
0          79        Y            1.0      7  
1          70        Y            1.0      8  
2          85        Y            1.0      7  
3          87        Y            1.0     10  
4          72        N            0.0      4

In [ ]:
# Drop several "unused" columns
vars = ["ID", "GenderGroup", "GlassesGroup", "CompleteGroup"]
df.drop(vars, axis=1, inplace = True)
df.head()

Age Gender Glasses  Height  Wingspan  CWDistance Complete  Score
0  56.0      F       Y    62.0      61.0          79        Y      7
1  26.0      F       Y    62.0      60.0          70        Y      8
2  33.0      F       Y    66.0      64.0          85        Y      7
3  39.0      F       N    64.0      63.0          87        Y     10
4  27.0      M       N    73.0      75.0          72        N      4

#Add and eliminate rows
In some cases it is requiered to add new observations (rows) to the data set

In [ ]:
# Print tail
df.tail()

Age Gender Glasses  Height  Wingspan  CWDistance Complete  Score
47  24.0      M       N    79.5      75.0          82        N      8
48  28.0      M       N    77.8      76.0          99        Y      9
49  30.0      F       N    74.6       NaN          71        Y      9
50   NaN      M       N    71.0      70.0         101        Y      8
51  27.0      M       N     NaN      71.5         103        Y     10

In [ ]:
df.loc[len(df.index)] = [26, 'F', 1, 66, 68, 3, 0, 'NaN']
df.tail()

Age Gender Glasses  Height  Wingspan  CWDistance Complete Score
48  28.0      M       N    77.8      76.0          99        Y     9
49  30.0      F       N    74.6       NaN          71        Y     9
50   NaN      M       N    71.0      70.0         101        Y     8
51  27.0      M       N     NaN      71.5         103        Y    10
52  26.0      F       1    66.0      68.0           3        0   NaN

In [ ]:
# Eliminate inserted row
df.drop([52], inplace = True )
df.tail()

Age Gender Glasses  Height  Wingspan  CWDistance Complete Score
47  24.0      M       N    79.5      75.0          82        N     8
48  28.0      M       N    77.8      76.0          99        Y     9
49  30.0      F       N    74.6       NaN          71        Y     9
50   NaN      M       N    71.0      70.0         101        Y     8
51  27.0      M       N     NaN      71.5         103        Y    10

# Cleaning your data: drop out unused columns and/or drop out rows with any missing values

In [ ]:
# Drop unused columns
#vars = ["ID", "GenderGroup", "GlassesGroup", "CompleteGroup"]
#df.drop(vars, axis=1, inplace = True)

vars = ["Age", "Gender", "Glasses", "Height", "Wingspan", "CWDistance", "Complete", "Score"]
df = df[vars]

# Drop rows with any missing values
#df = df.dropna()

# Drop unused columns and drop rows with any missing values
vars = ["Age", "Gender", "Glasses", "Height", "Wingspan", "CWDistance", "Complete", "Score"]
df = df[vars].dropna()

df

Age Gender Glasses  Height  Wingspan  CWDistance Complete Score
0   56.0      F       Y   62.00      61.0          79        Y     7
1   26.0      F       Y   62.00      60.0          70        Y     8
2   33.0      F       Y   66.00      64.0          85        Y     7
3   39.0      F       N   64.00      63.0          87        Y    10
4   27.0      M       N   73.00      75.0          72        N     4
5   24.0      M       N   75.00      71.0          81        N     3
6   28.0      M       N   75.00      76.0         107        Y    10
7   22.0      F       N   65.00      62.0          98        Y     9
8   29.0      M       Y   74.00      73.0         106        N     5
9   33.0      F       Y   63.00      60.0          65        Y     8
10  30.0      M       Y   69.50      66.0          96        Y     6
11  28.0      F       Y   62.75      58.0          79        Y    10
12  25.0      F       Y   65.00      64.5          92        Y     6
13  23.0      F       N   61.50      57.5          66        Y     4
14  31.0      M       Y   73.00      74.0          72        Y     9
15  26.0      M       Y   71.00      72.0         115        Y     6
16  26.0      F       N   61.50      59.5          90        N    10
17  27.0      M       N   66.00      66.0          74        Y     5
18  23.0      M       Y   70.00      69.0          64        Y     3
19  24.0      F       Y   68.00      66.0          85        Y     8
20  23.0      M       Y   69.00      67.0          66        N     2
21  29.0      M       N   71.00      70.0         101        Y     8
22  25.0      M       N   70.00      68.0          82        Y     4
23  26.0      M       N   69.00      71.0          63        Y     5
24  23.0      F       Y   65.00      63.0          67        N     3
25  28.0      M       N   75.00      76.0         111        Y    10
26  24.0      M       N   78.40      71.0          92        Y     7
27  25.0      M       Y   76.00      73.0         107        Y     8
29  38.0      F       Y   61.50      61.0          78        Y     7
30  27.0      F       Y   62.00      60.0          72        Y     8
31  33.0      F       Y   65.30      64.0          91        Y     7
32  38.0      F       N   64.00      63.0          86        Y    10
33  27.0      M       N   77.00      75.0         100        Y     8
34  24.0      F       N   67.80      62.0          98        Y     9
35  27.0      M       N   68.00      66.0          74        Y     5
36  25.0      F       Y   65.00      64.5          92        Y     6
37  26.0      F       N   61.50      59.5          90        Y     9
38  31.0      M       Y   73.00      74.0          72        Y     9
39  30.0      M       Y   69.50      66.0          96        Y     6
40  23.0      F       N   70.40      71.0          66        Y     4
41  26.0      M       Y   73.50      72.0         115        Y     6
42  28.0      F       Y   72.50      72.0          81        Y    10
43  26.0      F       Y   72.00      72.0          92        Y     8
44  30.0      F       Y   66.00      64.0          85        Y     7
45  39.0      F       N   64.00      63.0          87        Y    10
46  27.0      M       N   78.00      75.0          72        N     7
47  24.0      M       N   79.50      75.0          82        N     8
48  28.0      M       N   77.80      76.0          99        Y     9

#Activity: work with the iris dataset
Repeat this tutorial with the iris data set and respond to the following inquiries

Calculate the statistical summary for each quantitative variables. Explain the results

- Identify the name of each column
- Identify the type of each column
- Minimum, maximum, mean, average, median, standar deviation
Are there missing data? If so, create a new dataset containing only the rows with the non-missing data

Create a new dataset containing only the petal width and length and the type of Flower

Create a new dataset containing only the setal width and length and the type of Flower

Create a new dataset containing the setal width and length and the type of Flower encoded as a categorical numerical column

In [182]:
# Define the col names for the iris dataset
colnames = ["Sepal_Length", "Sepal_Width","Petal_Length","Petal_Width", "Flower"]

# Dataset url
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"

# Load the dataset from HHDD
dataset  = pd.read_csv(url, header = None, names = colnames )

dataset

Sepal_Length  Sepal_Width  Petal_Length  Petal_Width          Flower
0             5.1          3.5           1.4          0.2     Iris-setosa
1             4.9          3.0           1.4          0.2     Iris-setosa
2             4.7          3.2           1.3          0.2     Iris-setosa
3             4.6          3.1           1.5          0.2     Iris-setosa
4             5.0          3.6           1.4          0.2     Iris-setosa
..            ...          ...           ...          ...             ...
145           6.7          3.0           5.2          2.3  Iris-virginica
146           6.3          2.5           5.0          1.9  Iris-virginica
147           6.5          3.0           5.2          2.0  Iris-virginica
148           6.2          3.4           5.4          2.3  Iris-virginica
149           5.9          3.0           5.1          1.8  Iris-virginica

[150 rows x 5 columns]

In [183]:
# Print the number of rows
num_rows = len(dataset.index)
print(f"Number of rows: {num_rows}")

Number of rows: 150


In [184]:
# Print the number of columns
num_columns = len(dataset.columns)
print(f"Number of columns: {num_columns}")

Number of columns: 5


In [185]:
dataset.describe()

Sepal_Length  Sepal_Width  Petal_Length  Petal_Width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.054000      3.758667     1.198667
std        0.828066     0.433594      1.764420     0.763161
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

Las columnas nos indican los atributos de las observaciones (El largo y ancho de los sépalos de las flores, y el largo y ancho de los pétalos de las flores).



En la descripción de cada variable, se encuentra "count", que nos dice cuantas observaciones se hicieron para cada atributo (Sepal_Length,	Sepal_Width,	Petal_Length,	Petal_Width); "mean" nos dice la media de las 150 observaciones de cada atributo; "std" nos dice la desviación estándar o dispersión de datos; "min" nos dice el valor mínimo de cada atributo; "max" nos dice el valor máximo de cada atributo y las filas de "25%", "50%", "75%" nos indican los valores a estos porcentajes de las 150 observaciones.

In [189]:
# Create a new dataset containing only the petal width and length and the type of Flower:
df = dataset[["Petal_Length", "Petal_Width", "Flower"]]
df

Petal_Length  Petal_Width          Flower
0             1.4          0.2     Iris-setosa
1             1.4          0.2     Iris-setosa
2             1.3          0.2     Iris-setosa
3             1.5          0.2     Iris-setosa
4             1.4          0.2     Iris-setosa
..            ...          ...             ...
145           5.2          2.3  Iris-virginica
146           5.0          1.9  Iris-virginica
147           5.2          2.0  Iris-virginica
148           5.4          2.3  Iris-virginica
149           5.1          1.8  Iris-virginica

[150 rows x 3 columns]

In [190]:
# Create a new dataset containing only the setal width and length and the type of Flower:
df = dataset[["Sepal_Length", "Sepal_Width", "Flower"]]
df

Sepal_Length  Sepal_Width          Flower
0             5.1          3.5     Iris-setosa
1             4.9          3.0     Iris-setosa
2             4.7          3.2     Iris-setosa
3             4.6          3.1     Iris-setosa
4             5.0          3.6     Iris-setosa
..            ...          ...             ...
145           6.7          3.0  Iris-virginica
146           6.3          2.5  Iris-virginica
147           6.5          3.0  Iris-virginica
148           6.2          3.4  Iris-virginica
149           5.9          3.0  Iris-virginica

[150 rows x 3 columns]

In [197]:
# Create a new dataset containing the setal width and length and the type of Flower encoded as a categorical numerical column
df = dataset[["Sepal_Length", "Sepal_Width", "Flower"]]
df["Type_Flower_N"] = df.Flower.replace({'Iris-setosa': "1",'Iris-versicolor': "2",'Iris-virginica': "3"})
df

<ipython-input-197-d527b8616d6d>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Type_Flower_N"] = df.Flower.replace({'Iris-setosa': "1",'Iris-versicolor': "2",'Iris-virginica': "3"})


Sepal_Length  Sepal_Width          Flower Type_Flower_N
0             5.1          3.5     Iris-setosa             1
1             4.9          3.0     Iris-setosa             1
2             4.7          3.2     Iris-setosa             1
3             4.6          3.1     Iris-setosa             1
4             5.0          3.6     Iris-setosa             1
..            ...          ...             ...           ...
145           6.7          3.0  Iris-virginica             3
146           6.3          2.5  Iris-virginica             3
147           6.5          3.0  Iris-virginica             3
148           6.2          3.4  Iris-virginica             3
149           5.9          3.0  Iris-virginica             3

[150 rows x 4 columns]